fs_storage/models/fs_storage.py

```diff
@@ -0,0 +1,426 @@
+
+        fsspecs allows to nest filesystems. IOW, you can have a
+        filesystem that is based on another filesystem. The best example is
+        when you define a directory filesystem on top of a local filesystem.
+        (meaning that all files of the configured filesystem are stored in a
+        directory of the local filesystem). The root filesystem is the
+        filesystem that is not based on another filesystem.
+        """
+        self.ensure_one()
+        fs = self.fs
+        while hasattr(fs, "fs"):
+            fs = fs.fs
+        return fs
+
+    def _get_filesystem_storage_path(self) -> str:
+        """Get the path to the storage directory.
+
+        This path is relative to the odoo filestore.and is used as root path
+        when the protocol is filesystem.
+        """
+        self.ensure_one()
+        path = os.path.join(self.env["ir.attachment"]._filestore(), "storage")
+        if not os.path.exists(path):
+            os.makedirs(path)
+        return path
+
+    @property
+    def _odoo_storage_path(self) -> str:
+        """Get the path to the storage directory.
+
+        This path is relative to the odoo filestore.and is used as root path
+        when the protocol is filesystem.
+        """
+        if not self.__odoo_storage_path:
+            self.__odoo_storage_path = self._get_filesystem_storage_path()
+        return self.__odoo_storage_path
+
+    def _recursive_add_odoo_storage_path(self, options: dict) -> dict:
+        """Add the odoo storage path to the options.
+
+        This is a recursive function that will add the odoo_storage_path
+        option to the nested target_options if the target_protocol is
+        odoofs
+        """
+        if "target_protocol" in options:
+            target_options = options.get("target_options", {})
+            if options["target_protocol"] == "odoofs":
+                target_options["odoo_storage_path"] = self._odoo_storage_path
+                options["target_options"] = target_options
+            self._recursive_add_odoo_storage_path(target_options)
+        return options
+
+    def _get_filesystem(self) -> fsspec.AbstractFileSystem:
+        """Get the fsspec filesystem for this backend.
+
+        See https://filesystem-spec.readthedocs.io/en/latest/api.html
+        #fsspec.spec.AbstractFileSystem
+
+        :return: fsspec.AbstractFileSystem
+        """
+        self.ensure_one()
+        options = self.json_options
+        if self.protocol == "odoofs":
+            options["odoo_storage_path"] = self._odoo_storage_path
+        options = self._recursive_add_odoo_storage_path(options)
+        fs = fsspec.filesystem(self.protocol, **options)
+        directory_path = self.directory_path
+        if directory_path:
+            fs = fsspec.filesystem("rooted_dir", path=directory_path, fs=fs)
+        return fs
+
+    @deprecated("Please use _get_filesystem() instead and the fsspec API directly.")
```

Comment:
It is not clear for me why this method and the other after are deprecated.
These methods are helpers that could be used in several submodules.
Of course each submodule could re-implement its own method taht will 
- decode data
- create subdirectories if missing
- create the file
But why not keeping it here to avoid code duplication in those submodules ?
I guess I am missing something... @lmignon

Reply:
@florian-dacosta The API provided by the `fsspec`lib is more complete. I've the feeling that in a lot of cases, you could find into this lib, methods more suitable to your use cases.  I would like to avoid to have to maintain yet an other one api to manager a file system storage than the one provided by the lib.